### PR TITLE
chore(pipeline-storage): inject basepath as dependency

### PR DESCRIPTION
**Pull request by mabaasit**
In this PR, we aligned PipelineStorage with other storage implementations by injecting basepath as a dependency.

### Checklist
- [x] New tests and/or benchmarks are included
- [ ] Documentation is changed or added
- [ ] I have signed the MongoDB Contributor License Agreement (https://www.mongodb.com/legal/contributor-agreement)

## Motivation and Context

- [ ] Bugfix
- [ ] New feature
- [ ] Dependency update
- [ ] Misc

## Types of changes

- [ ] *Backport Needed*
- [ ] Patch (non-breaking change which fixes an issue)
- [x] Minor (non-breaking change which adds functionality)
- [ ] Major (fix or feature that would cause existing functionality to change)

### CODE REVIEW

File: packages/compass-aggregations/src/utils/pipeline-storage.spec.js [hunk 1]
```diff
@@ -1,35 +1,28 @@
...
 
-const pipelineStorage = new PipelineStorage();
-
 describe('PipelineStorage', function () {
   let tmpDir;
+  let pipelineStorage;
   beforeEach(function () {
-    tmpDir = fs.mkdtempSync(path.join(os.tmpdir(), 'SavedPipelines'));
+    tmpDir = fs.mkdtempSync(os.tmpdir());
     fs.mkdirSync(path.join(tmpDir, 'SavedPipelines'));
```
Comment (mcasimir) [inline packages/compass-aggregations/src/utils/pipeline-storage.spec.js:20]: I would expect ensuring the path exists being done by the storages

File: packages/compass-aggregations/src/stores/store.ts [hunk 2]
```diff
@@ -156,7 +159,7 @@ const configureStore = (options: ConfigureStoreOptions) => {
     options.dataProvider.dataProvider as DataService,
     initialPipelineSource
   );
-  const pipelineStorage = new PipelineStorage();
+  const pipelineStorage = new PipelineStorage(basepath);
```
Comment (mcasimir) [inline packages/compass-aggregations/src/stores/store.ts:162]: why is this a better alternative than having it inside the storages? if the reason is testing you can still have a default storage path and have the constructor accept a param to override it.

in this way you are giving a "passing knowledge" to `store.ts` of where the files are saved, which store.ts can't use directly.

File: packages/compass-aggregations/src/utils/pipeline-storage.spec.js (hunk 1 above)
Comment (mabaasit) [reply packages/compass-aggregations/src/utils/pipeline-storage.spec.js:20]: I was thinking of doing all the directory setup as the app starts in the main process somewhere [here](https://github.com/mongodb-js/compass/blob/main/packages/compass/src/main/application.ts#L60). That way we will ensure we have all the necessary directories setup when we try to read from them.

File: packages/compass-aggregations/src/stores/store.ts (hunk 2 above)
Comment (mabaasit) [reply packages/compass-aggregations/src/stores/store.ts:162]: Yes, I did it for that reason (testability). I do see your point about sharing knowledge in `store.ts`, so I don't have a strong preference here.

File: packages/compass-aggregations/src/utils/pipeline-storage.spec.js (hunk 1 above)
Comment (mcasimir) [reply packages/compass-aggregations/src/utils/pipeline-storage.spec.js:20]: nice idea! but similar to the other comment, it seems that storages have the info on what's the folder: https://github.com/mongodb-js/compass/pull/4674/files/e7050dd7b6e504b4840c150e6cde7b52d0a4d1b1#diff-cda704877ab6cd60ae17d533f713bd373be3a3458105773b5a06e290f25e997fR68

So the storage would still be the "information expert" here. By requiring `_init` to run `mkdir` we would need to give the `_init` code a piece of knowledge that storages already have, and introduce a stronger coupling: where we want to save files and how (on a filesystem vs somewhere else). 

We could consider adding a `.setup()` method instead and call that in the init code .. that way we would have a fail-fast type of experience where if is not possible to write on a directory .. compass would crash??

Or, to get rid of the coupling entirely .. just use a lazy initialization, and have mkdir being called once if it wasn't before. I would prefer this way if that's what is happening right now, so we don't incur in the risk of having new issues that we didn't have before.

File: packages/compass-aggregations/src/stores/store.ts (hunk 2 above)
Comment (mcasimir) [reply packages/compass-aggregations/src/stores/store.ts:162]: Yeah, btw having a way to change the base directory in the constructor is a great idea, we should do it! If you are ok with it, let's also keep the default in the storages so we make it both testable and we won't leak details in the client code.

**Comment by mabaasit** [comment]
I am closing this PR as we agreed to:
1. Have optional `options` param across storage classes where we will have `basepath`
2. Ensure storage folder (PipelineStorage, FavoriteQuery ...) exist in the storage class (this avoids weird setup in tests as well)
3. Refactor common code to access files

**Pull request closed without merging**